File: src/examples/protect-list/pre-filter-fine.go
package main

import (
	"context"
	"fmt"
	"io"

	v1beta2 "github.com/project-kessel/kessel-sdk-go/kessel/inventory/v1beta2"
	v2 "github.com/project-kessel/kessel-sdk-go/kessel/rbac/v2"
)

//#region setup
// Initialize Kessel client (see "Protect an endpoint" guide for full setup)
func setupClient() (v1beta2.KesselInventoryServiceClient, error) {
	inventoryClient, _, err := v1beta2.NewClientBuilder("localhost:9000").
		Insecure().
		Build()
	if err != nil {
		return nil, fmt.Errorf("failed to create client: %w", err)
	}
	return inventoryClient, nil
}

//#endregion

//#region get-resource-ids
// getAccessibleResourceIDs gets all resource IDs the user can access with the given permission.
//
// This uses StreamedListObjects with your specific resource type to find
// exactly which resource IDs the user can access.
func getAccessibleResourceIDs(
	ctx context.Context,
	client v1beta2.KesselInventoryServiceClient,
	userID string,
	permission string,
) ([]string, error) {
    subject := v2.PrincipalSubject(userID, "redhat")
	reporterType := "myservice"
	resourceType := &v1beta2.RepresentationType{
		ResourceType: "integration",
		ReporterType: &reporterType,
	}

	request := &v1beta2.StreamedListObjectsRequest{
		ObjectType: resourceType,
		Relation:   permission,
		Subject:    subject,
	}

	stream, err := client.StreamedListObjects(ctx, request)
	if err != nil {
		return nil, fmt.Errorf("failed to start stream: %w", err)
	}

	// Get accessible resource IDs
	var resourceIDs []string
	for {
		response, err := stream.Recv()
		if err == io.EOF {
			break
		}
		if err != nil {
			return nil, fmt.Errorf("error receiving from stream: %w", err)
		}
		resourceIDs = append(resourceIDs, response.Object.ResourceId)
	}

	return resourceIDs, nil
}

//#endregion

//#region filter-query
// Integration represents an integration in your database
type Integration struct {
	ID          string
	Name        string
	WorkspaceID string
}

// Database is a placeholder for your database interface
type Database interface {
	Query(query string, params ...interface{}) ([]*Integration, error)
}

// listIntegrations returns integrations the user can access.
//
// Pre-filtering (fine-grained): Get allowed resource IDs first, then query only those.
func listIntegrations(ctx context.Context, client v1beta2.KesselInventoryServiceClient, db Database, userID string) ([]*Integration, error) {
	// Step 1: Get resource IDs the user can access
	allowedIDs, err := getAccessibleResourceIDs(ctx, client, userID, "view")
	if err != nil {
		return nil, err
	}

	if len(allowedIDs) == 0 {
		return []*Integration{}, nil
	}

	// Step 2: Query database with resource ID filter
	query := "SELECT id, name, workspace_id FROM integrations WHERE id = ANY($1)"
	integrations, err := db.Query(query, allowedIDs)
	if err != nil {
		return nil, fmt.Errorf("failed to query integrations: %w", err)
	}

	return integrations, nil
}

//#endregion

File: src/examples/protect-list/pre-filter-fine.ts
import { ClientBuilder } from "@project-kessel/kessel-sdk/kessel/inventory/v1beta2";
import { StreamedListObjectsRequest } from "@project-kessel/kessel-sdk/kessel/inventory/v1beta2/streamed_list_objects_request";
import { RepresentationType } from "@project-kessel/kessel-sdk/kessel/inventory/v1beta2/representation_type";
import { principalSubject } from "@project-kessel/kessel-sdk/kessel/rbac/v2";

//#region setup
// Initialize Kessel client (see "Protect an endpoint" guide for full setup)
const client = new ClientBuilder("localhost:9000")
  .insecure()
  .buildAsync();
//#endregion

//#region get-resource-ids
/**
 * Get all resource IDs the user can access with the given permission.
 *
 * This uses StreamedListObjects with your specific resource type to find
 * exactly which resource IDs the user can access.
 */
async function getAccessibleResourceIds(userId: string, permission: string): Promise<string[]> {
  const subject = principalSubject(userId, "redhat");
  const resourceType: RepresentationType = {
    resourceType: "integration",
    reporterType: "myservice"
  };

  const request: StreamedListObjectsRequest = {
    objectType: resourceType,
    relation: permission,
    subject: subject
  };

  // Get accessible resource IDs
  const resourceIds: string[] = [];
  for await (const response of client.streamedListObjects(request)) {
    if (response.object?.resourceId) {
      resourceIds.push(response.object.resourceId);
    }
  }

  return resourceIds;
}
//#endregion

//#region filter-query
/**
 * List integrations the user can access.
 *
 * Pre-filtering (fine-grained): Get allowed resource IDs first, then query only those.
 */
async function listIntegrations(db: Database, userId: string): Promise<Integration[]> {
  // Step 1: Get resource IDs the user can access
  const allowedIds = await getAccessibleResourceIds(userId, "view");

  if (allowedIds.length === 0) {
    return [];
  }

  // Step 2: Query database with resource ID filter
  const integrations = await db.query(
    "SELECT * FROM integrations WHERE id = ANY($1)",
    [allowedIds]
  );

  return integrations;
}

// Database interface placeholder
interface Database {
  query(sql: string, params: any[]): Promise<Integration[]>;
}

interface Integration {
  id: string;
  name: string;
  workspaceId: string;
}
//#endregion
